### PR TITLE
Unbind project and workspace directories

### LINKED ISSUE #1030: Unbind project and workspace directories
### Rationale

We have a plan to unify xodballs and multifile project and use the native OS dialog to save/load both kinds. In this scenario, it would be very hard to tell a xoder his multifile project must be an immediate child of the workspace directory. It would potentially lead to an infinite amount of wtfs.

So we can allow placing a multifile project anywhere on FS. We need libraries and they are bound to a workspace. OK, let make the workspace a distinct concept. It can be anywhere and by default is at `$HOME/xod`. Such behavior is something like Photoshop does for storing brushes, swatches, etc: a user can change this dir in preferences, but for an average user it does not matter much.

### Acceptance criteria

- [x] I can copy a multifile XOD project anywhere on FS, double-click the `project.xod` and it will open correctly in XOD
- [x] The last active workspace directory will be used for libraries
- [x] By default the workspace directory is `$HOME/xod`
- [x] I can switch the workspace directory like I do already: “File → Select Workspace”¹
- [x] `xodc` supports `--workspace` flag where appropriate, which accepts directory path
- [x] `xodc` uses `XOD_WORKSPACE` env var if `--workspace` is not passed
- [x] `xodc` assumes `$HOME/xod` as the last fallback

¹ Suggest renaming to “File → Switch Workspace” for consistent terminology.

### PR DESCRIPTION
**Pull request by brusherru**
It closes #1030.
If you run IDE by opening any of associated files — you can get popup "Please select Project" sometimes (sometimes not, there is a racing condition now). But I prefer to avoid writing temporary kludge, cause it will be removed later in the #1036.

### CODE REVIEW

File: packages/xod-client-electron/src/app/utils.js [hunk 1]
```diff
@@ -12,3 +15,42 @@ export const errorToPlainObject = R.when(
     R.identity,
   ])
 );
+
+// :: String -> Maybe Path
+const maybePath = R.ifElse(
+  path.isAbsolute.bind(path),
```
Comment (nkrkv) [inline packages/xod-client-electron/src/app/utils.js]: The `maybePath` name is misguiding. Why we ever need a voodoo dance with `isAbsolute`? I’m sure there is a reason, but I don’t understand which one. A comment will help.

File: packages/xod-client-electron/src/app/utils.js [hunk 2]
```diff
@@ -12,3 +15,42 @@ export const errorToPlainObject = R.when(
...
+ * It accepts an app and returns a getter function,
+ * that will return `Promise Path Null`.
+ */
+// :: App -> (_ -> Promise Path Null)
+export const getFilePathToOpen = (app) => {
+  // Windows & Linux
+  let pathToOpen = R.compose(
+    R.chain(maybePath),
+    Maybe
+  )(process.argv[1]);
```
Comment (nkrkv) [inline packages/xod-client-electron/src/app/utils.js]: Isn’t `process.argv[1]` fragile?

- `electron . /path/to/project`
- `node /path/to/electron /path/to/xod /path/to/project`
- `xod-client-electron /path/to/project`

Does `yarn start:electron -- -- /path/to/project` work on your machine?

**Comment by brusherru** [comment]
@nkrkv check it out, please :)

File: packages/xod-cli/src/xodc-install.js [hunk 3]
```diff
@@ -3,17 +3,20 @@ import * as xodFs from 'xod-fs';
...
  * @return {Promise.<void>}
  */
-export default function install(swaggerUrl, libQuery, distPath) {
+export default function install(swaggerUrl, libQuery, workspace) {
   return Promise.all([
     fetchLibData(swaggerUrl, libQuery),
-    xodFs.findClosestWorkspaceDir(distPath),
+    xodFs.findClosestWorkspaceDir(
+      getWorkspacePath(workspace)
+    ),
```
Comment (evgenykochetkov) [inline packages/xod-cli/src/xodc-install.js]: So we pass a `Path` here, which is defaulted to a default workspace path if it's `null`, and then scan that path and it's parent directories for `.xodworkspace`🤔
Do we really need it? I think something like `R.compose(ensureThatItsAWorkspace, getWorkspacePath)(workspace)` should be enough.

File: packages/xod-fs/src/save.js [hunk 4]
```diff
@@ -71,7 +71,8 @@ export const saveArrangedFiles = R.curry((pathToWorkspace, data) => {
 
 // :: Path -> Project -> Promise Project Error
 export const saveProjectEntirely = R.curry(
-  (workspacePath, project) => Promise.resolve(project)
+  (distPath, project) => Promise.resolve(project)
+    // .then(R.tap(a => console.dir(a, { depth: 2, colors: true })))
```
Comment (evgenykochetkov) [inline packages/xod-fs/src/save.js]: Sneaky commented out console.log

File: packages/xod-fs/test/welcomeToXod.spec.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ describe('welcome-to-xod', () => {
   const projectPath = 'welcome-to-xod';
 
   it('should load as a valid project', () =>
-    loadProjectWithLibs([], projectPath, workspace)
+    loadProjectWithLibs([workspace], path.resolve(workspace, projectPath))
```
Comment (evgenykochetkov) [inline packages/xod-fs/test/welcomeToXod.spec.js]: A test with multiple workspaces would be awesome(for example, finding one library in the first workspace, and another in the other one, or "preferring" a version of a library which exists in both workspaces from first workspace, etc), but I am not insisting 😁

Comment (brusherru) [reply packages/xod-fs/test/welcomeToXod.spec.js]: Added one, but into `loadLib.spec.js`

**Comment by brusherru** [comment]
@nkrkv check it out, please

File: packages/xod-fs/src/save.js [hunk 6]
```diff
@@ -71,7 +71,7 @@ export const saveArrangedFiles = R.curry((pathToWorkspace, data) => {
 
 // :: Path -> Project -> Promise Project Error
 export const saveProjectEntirely = R.curry(
-  (workspacePath, project) => Promise.resolve(project)
+  (distPath, project) => Promise.resolve(project)
```
Comment (nkrkv) [inline packages/xod-fs/src/save.js]: `dirPath` ?

File: packages/xod-cli/src/utils.js [hunk 7]
```diff
@@ -16,4 +17,12 @@ export const runCommand = R.uncurryN(2, options =>
   )
 );
 
+// :: Nullable Path -> Path
+export const getWorkspacePath = R.compose(
+  expandHomeDir,
+  R.defaultTo(
+    process.env.XOD_WORKSPACE || DEFAULT_WORKSPACE_PATH
```
Comment (nkrkv) [inline packages/xod-cli/src/utils.js]: I found that:

```js
export const DEFAULT_WORKSPACE_PATH = '~/xod/';
```

From my experience keeping trailing slashes while defining dirs lead to huge spaghetti in the future. As a rule of thumb: never end a path component with a slash, let the consumers add one if they want.

Can we fix it?

Comment (brusherru) [reply packages/xod-cli/src/utils.js]: Done

**Pull request merged**